nuxeo-runtime/nuxeo-stream/src/main/java/org/nuxeo/lib/stream/computation/log/ComputationRunner.java

```diff
@@ -174,51 +188,70 @@ public boolean waitForAssignments(Duration timeout) throws InterruptedException
     @Override
     public void run() {
         threadName = Thread.currentThread().getName();
-        boolean interrupted = false;
-        boolean normalTermination = false;
         computation = supplier.get();
         log.debug(metadata.name() + ": Init");
         registerMetrics();
+        ReturnCode returnCode = ReturnCode.TERMINATE;
+        computation.init(context);
+        log.debug(metadata.name() + ": Start");
         try {
-            computation.init(context);
-            log.debug(metadata.name() + ": Start");
-            processLoop();
-            normalTermination = true;
+            int i = 0;
+            do {
+                returnCode = runOnce();
+                if (ReturnCode.CHECKPOINT_ERROR.equals(returnCode)) {
+                    i++;
+                    if (i >= CHECKPOINT_MAX_RETRY) {
+                        log.error(metadata.name() + ": Terminate computation because too many checkpoint errors");
+                    } else {
+                        log.warn("Wait a bit after checkpoint error, retry #" + i);
+                        Thread.sleep(CHECKPOINT_PAUSE_MS);
+                    }
+                }
+            } while (ReturnCode.CHECKPOINT_ERROR.equals(returnCode) && i < CHECKPOINT_MAX_RETRY);
         } catch (InterruptedException e) {
-            interrupted = true; // Thread.currentThread().interrupt() in finally
-            // this is expected when the pool is shutdownNow
-            String msg = metadata.name() + ": Interrupted";
-            if (log.isTraceEnabled()) {
-                log.debug(msg, e);
-            } else {
-                log.debug(msg);
-            }
-        } catch (Exception e) {
-            if (Thread.currentThread().isInterrupted()) {
-                // this can happen when pool is shutdownNow throwing ClosedByInterruptException
-                log.info(metadata.name() + ": Interrupted", e);
-            } else {
-                log.error(metadata.name() + ": Exception in processLoop: " + e.getMessage(), e);
-                throw e;
-            }
+            returnCode = ReturnCode.INTERRUPTED;
         } finally {
             try {
                 computation.destroy();
                 closeTailer();
             } finally {
-                if (interrupted) {
+                if (ReturnCode.INTERRUPTED.equals(returnCode)) {
                     Thread.currentThread().interrupt();
                 }
             }
-            if (normalTermination || interrupted) {
-                log.debug(metadata.name() + ": Terminated");
+        }
+    }
+
+    protected ReturnCode runOnce() {
+        try {
+            processLoop();
+        } catch (CheckPointException e) {
+            log.error(metadata.name() + ": Error during checkpoint, processing will be duplicated: " + e.getMessage(),
```

Comment:
As it is an expected failure, shouldn't we log at warn level?

Reply:
The `e.getMessage()` will also contains the metadata name, don't we want to remove it from the exception in `checkpointIfNecessary` method?

Reply:
I think the error level is ok because there is a risk of duplicate processing which can be a problem if the processing is not idempotent.
for the message the format is now unified using the computation name as a prefix to ease searching in log,